File: PRACTICAS-PDOO/IRRGARTEN/Versiones_en_Java_Ruby/PROYECTO1-JAVA/src/irrgarten/Dice.java

package irrgarten;

import java.util.Random;

/**
 *
 * @author angel_rodriguez
 */

/* 
    Como no podemos crear instancias de esta clase (Dice), no tiene sentido que 
    los métodos sean de instancia por lo que tendremos que poner métodos de 
    clase. En Java, se ponen como static 
*/

/* Esta clase tiene la responsabilidad de tomar todas las decisiones que 
   dependen del azar en el juego. Es como una especie de dado, pero algo más 
   sofisticado, ya que no proporciona simplemente un número del 1 al 6, sino 
   decisiones concretas en base a una serie de probabilidades establecidas.
*/

public class Dice {
    
    // Atributos de clase privados. 
    
        // private static final significan que son CONSTANTES de clase y privadas.
    
    private static final int MAX_USES = 5;              //(número máximo de usos de armas y escudos)
    private static final float MAX_INTELLIGENCE = (float) 10.0;  //(valor máximo para la inteligencia de jugadores y monstruos)
    private static final float MAX_STRENGTH = (float) 10.0;      //(valor máximo para la fuerza de jugadores y monstruos)
    private static final float RESURRECT_PROB = (float) 0.3;    //(probabilidad de que un jugador sea resucitado en cada turno)
    private static final int WEAPONS_REWARD = 2;        //(numero máximo de armas recibidas al ganar un combate)
    private static final int SHIELDS_REWARD = 3;        //(numero máximo de escudos recibidos al ganar un combate)
    private static final int HEALTH_REWARD = 5;         //(numero máximo de unidades de salud recibidas al ganar un combate)
    private static final int MAX_ATTACK = 3;            //(máxima potencia de las armas)
    private static final int MAX_SHIELD = 2;            //(máxima potencia de los escudos)
    private static final Random generator = new Random();
    
    // Métodos de clase públicos.
    
    public static int randomPos(int max){
        int randomPos = 0;
        randomPos = generator.nextInt(max);
        return randomPos;
    }
    
    public static int whoStarts(int nplayers){
        int randomStart = 0;
        randomStart = generator.nextInt(nplayers);
        return randomStart;
    }
    
    /* 
    - El método Random.nextInt() devuelve un entero aleatorio.
      Si le damos un parámetro, nextInt(máx), y máx es de tipo int, nos 
      devolverá un entero. ESTO PRODUCE DESDE 0 A MÁX-1.
      Si queremos que nos produzca hasta max (inclusive), tenemos que poner (máx+1)
      Si lo queremos poner en rangos, por ejemplo desde 1 a 10 incluidos, tengo
      que ponerlo como (nextInt(max) + 1)
    
    - El método Random.netxFloat no aceptan parámetros, por lo que nos
      devolverá un numero aleatorio entre 0,0 y 1,0, y eso tenemos que 
      sumarlo a un entero.
    */

    public static float randomIntelligence(){
        float randomIntelligence = (float) 0.0;
        int randomNum = 0;
        
        do{
            randomNum = generator.nextInt((int) MAX_INTELLIGENCE);
            randomIntelligence = (generator.nextFloat() + randomNum);
        } while (randomIntelligence > MAX_INTELLIGENCE);
        
        return randomIntelligence;
    }
    
    public static float randomStrength(){
        float randomStrength = (float) 0.0;
        int randomNum = 0;
        
        do{
            randomNum = generator.nextInt( (int) MAX_STRENGTH );
            randomStrength = (generator.nextFloat() + randomNum);
        } while (randomStrength > MAX_STRENGTH);
        
        return randomStrength;
    }
    
    public static boolean resurrectPlayers(){
        float numRandom = generator.nextFloat();
        if(numRandom <= RESURRECT_PROB){
            return true;
        }else{
            return false;
        }
    }
    
    public static int weaponsReward(){
        int weaponsReward = 0;
        weaponsReward = generator.nextInt(WEAPONS_REWARD + 1);
        return weaponsReward;
        
    }
    
    public static int shieldsReward(){
        int shieldsReward = 0;
        shieldsReward = generator.nextInt(SHIELDS_REWARD + 1);
        return shieldsReward;
    }
   
    public static int healthReward(){
        int healthReward = 0;
        healthReward = generator.nextInt(HEALTH_REWARD + 1);
        return healthReward;
    }
   
    public static float weaponPower(){
        float weaponPower = (float) 0.0;
        int randomNum = 0;
        
        do{
            randomNum = generator.nextInt(MAX_ATTACK);
            weaponPower = (generator.nextFloat() + randomNum);
        } while (weaponPower >= MAX_INTELLIGENCE);
        
        return weaponPower;
    }

    public static float shieldPower(){
        float shieldPower = (float) 0.0;
        int randomNum = 0;
        
        do{
            randomNum = generator.nextInt(MAX_SHIELD);
            shieldPower = (generator.nextFloat() + randomNum);
        } while (shieldPower >= MAX_SHIELD);
        
        return shieldPower;
    }
    
    public static int usesLeft(){
        int randomUses = 0;
        randomUses = generator.nextInt(MAX_USES + 1);
        return randomUses;
    }
    
    public static float intensity(float competence){
        float intensity = (float) 0.0;
        int randomNum = 0;
        
        do{
            randomNum = generator.nextInt((int) competence);
            intensity = (generator.nextFloat() + randomNum);
        } while (intensity >= competence);
        
        return intensity;       
    }
    
    public static boolean discardElement (int usesLeft){
        // Caso extremo: si el número de usos es el máximo, devuelve false
        if (usesLeft == MAX_USES) {
            return false;
        }
        
        // Caso extremo: si el número de usos es 0, devuelve true
        if (usesLeft == 0) {
            return true;
        }

        // Calcula la probabilidad inversamente proporcional
        double probabilidad = 1.0 / (usesLeft + 1);

        // Genera un número aleatorio entre 0 y 1
        double randomValue = generator.nextDouble();

        // Compara el número aleatorio con la probabilidad
        return randomValue < probabilidad;
    } 

}

File: PRACTICAS-PDOO/IRRGARTEN/Versiones_en_Java_Ruby/PROYECTO3-RUBY/Irrgarten/dice.rb
# frozen_string_literal: true
#encoding:utf-8

# File: dice.rb
# Author: Ángel Rodríguez Faya

module Irrgarten

  # Esta clase tiene la responsabilidad de tomar todas las decisiones que dependen del azar en el juego.
  # Es como una especie de dado, pero algo más sofisticado, ya que no proporciona simplemente un
  # número del 1 al 6, sino decisiones concretas en base a una serie de probabilidades establecidas.

  #Como no podemos crear instancias de esta clase (Dice), no tiene sentido que los métodos sean de instancia
  # por lo que tendremos que poner métodos de clase. En Ruby, se ponen como self._______

  class Dice
    # Atributos de clase privados
    @@MAX_USES = 5              #(número máximo de usos de armas y escudos)
    @@MAX_INTELLIGENCE = 10.0   #(valor máximo para la inteligencia de jugadores y monstruos)
    @@MAX_STRENGTH = 10.0       #(valor máximo para la fuerza de jugadores y monstruos)
    @@RESURRECT_PROB = 0.3      #(probabilidad de que un jugador sea resucitado en cada turno)
    @@WEAPONS_REWARD = 2        #(numero máximo de armas recibidas al ganar un combate)
    @@SHIELDS_REWARD = 3        #(numero máximo de escudos recibidos al ganar un combate)
    @@HEALTH_REWARD = 5         #(numero máximo de unidades de salud recibidas al ganar un combate)
    @@MAX_ATTACK = 3            #(máxima potencia de las armas)
    @@MAX_SHIELD = 2            #(máxima potencia de los escudos)
    @@generator = Random.new

    # Métodos de clase públicos

    # El método rand() de la clase Random devuelve un float aleatorio si no le damos parámetros.
    # Si le damos un parámetro, rand(máx), y máx es de tipo int, nos devolverá un entero. ESTO PRODUCE DESDE 0 A MÁX-1.
    # Si queremos que nos produzca hasta max (inclusive), tenemos que poner rand(máx+1)
    # También lo podemos poner en rangos, por ejemplo: rand(0..10)

    public

    def self.random_pos(max)
      @@generator.rand(max.to_i)
    end

    def self.who_starts(n_players)
      @@generator.rand(n_players.to_i)
    end

    def self.random_intelligence
      @@generator.rand(@@MAX_INTELLIGENCE)
    end

    def self.random_strength
      @@generator.rand(@@MAX_STRENGTH)
    end

    def self.resurrect_player
      random_float = @@generator.rand
      if random_float <= @@RESURRECT_PROB
        return true
      else
        return false
      end
    end

    def self.weapons_reward
      @@generator.rand(@@WEAPONS_REWARD + 1)
    end

    def self.shields_reward
      @@generator.rand(@@SHIELDS_REWARD + 1)
    end

    def self.health_reward
      @@generator.rand(@@HEALTH_REWARD + 1)
    end

    def self.weapon_power
      @@generator.rand(@@MAX_ATTACK)
    end

    def self.shields_power
      @@generator.rand(@@MAX_SHIELD)
    end

    def self.uses_left
      @@generator.rand(@@MAX_USES + 1)
    end

    def self.intensity(competence)
      @@generator.rand(competence.to_f)
    end

    def self.discard_element(uses_left)
      uses_left = uses_left.to_i
      # Caso extremo: si el número de usos es el máximo, devuelve false
      return false if uses_left == @@MAX_USES

      # Caso extremo: si el número de usos es 0, devuelve true
      return true if uses_left == 0

      # Calcula la probabilidad inversamente proporcional
      probabilidad = 1.0/(uses_left.to_f + 1.0)

      # Genera un número aleatorio entre 0 y 1
      random_value = @@generator.rand

      # Compara el número aleatorio con la probabilidad
      return random_value < probabilidad

    end

  end   # end de la clase Dice

end  # end del módulo Irrgarten